source/server/listener_impl.h

```diff
@@ -204,6 +208,51 @@ class PerListenerFactoryContextImpl : public Configuration::ListenerFactoryConte
   ListenerImpl& listener_impl_;
 };
 
+struct FilterChainRebuildInfo {
+  FilterChainRebuildInfo(
+      const std::string& worker_name, const std::string& listener_name,
+      const envoy::config::listener::v3::FilterChain* const& filter_chain_message)
+      : worker_name_(worker_name), listener_name_(listener_name),
+        filter_chain_message_(filter_chain_message) {}
+
+  const std::string& worker_name_;
+  const std::string& listener_name_;
+  const envoy::config::listener::v3::FilterChain* const& filter_chain_message_;
+};
+
+using FilterChainRebuildInfoPtr = std::unique_ptr<FilterChainRebuildInfo>;
+
+class FilterChainRebuilder {
+public:
+  FilterChainRebuilder(ListenerImpl& listener,
+                       const envoy::config::listener::v3::FilterChain* const& filter_chain);
+  ~FilterChainRebuilder();
+
+  void addRebuildInfo(FilterChainRebuildInfoPtr rebuild_info);
+  bool isCompleted() { return rebuild_complete_; }
+  Init::Manager& initManager() { return *rebuild_init_manager_; }
+  void startRebuilding() { rebuild_init_manager_->initialize(rebuild_watcher_); }
+  void callbackToWorkers();
```

Comment:
Each filter chain message is assigned to only one rebuilder, which includes these necessary methods.